src/crystal/hasher.cr

```diff
@@ -1,18 +1,44 @@
 require "random/secure"
 
+# TODO: use flag?(:bits64) for Crystal > 0.23.1
+{% if flag?(:x86_64) || flag?(:aarch64) %}
+  require "digest/siphash"
+{% else %}
+  require "digest/halfsiphash"
+{% end %}
+
 # :nodoc:
 struct Crystal::Hasher
   # Implementation of a Hasher to compute a fast and safe hash
   # value for primitive and basic Crystal objects. All other
   # hashes are computed based on these.
   #
-  # TODO: the implementation is naive and should probably use
-  # another algorithm like SipHash or FNV.
+  # Relies on the SipHash family of pseudo random functions. Since we never
+  # disclose the result of the hashes, we don't need the cryptographically
+  # verified siphash-2-4, but can use the faster siphash-1-3 alternative.
+  #
+  # On 32-bit systems, we prefer the halfsiphash-1-3 alternative (32-bit hashes)
+  # that should perform better than siphash-1-3 (64-bit hashes).
+
+  # TODO: use flag?(:bits64) for Crystal > 0.23.1
+  {% if flag?(:x86_64) || flag?(:aarch64) %}
+    alias SipHash = Digest::SipHash
+    alias HashType = UInt64
+  {% else %}
+    alias SipHash = Digest::HalfSipHash
+    alias HashType = UInt32
```

Comment:
I’m very unhappy to handle two different hash types on different architectures. Is it really unavoidable?


I should note that you must update StringPool accordingly in this case